jax/numpy/lax_numpy.py

```diff
@@ -3140,6 +3140,7 @@ def unique(ar, return_index=False, return_inverse=False,
 
 ### Indexing
 
+@partial(jit, static_argnums=(1,))
```

Comment:
Will this recompile for every new value of `idx` now? I'm not sure we want that...

Reply:
Seems like Peter tried jitting `_gather` before (below) but the problem is that even there we have (I think?) arguments which are sometimes static and sometimes dynamic (in the sense of, sometimes we want to recompile when they change and sometimes we don't) depending on the indexing mode. I am...not sure how to solve this.

Reply:
I think it's not to do this, since it'd mean recompiling here:

```python
idx1 = np.array([0, 1, 2])
idx2 = np.array([1, 2, 3])

x = np.arange(10)
x[idx1]  # compiles
x[idx2]  # no recompile
```

We could do something like tree-flatten the idx, then cache on equality of tree structures and the shaped avals of leaves (that is, the tree structure could be a static arg).

But in any case, I'd like to get @hawkinsp's insights since he's thought about both `_unstack` (the previous path) and `jit`ting indexing logic. The SDA case may be more restricted than general gathers, so perhaps a specialized path like `_unstack` still makes sense.

Reply:
This makes me think #2813 is a good idea.

Reply:
Yes, a `jit` here is not a good idea.

The slightly better idea sketched out here already is to split the indices into a static part and a dynamic part (`_split_index_for_jit`). We then call a helper function (`_gather`) that is `jit`-ted with only the static part being a `static_argnum`. Unfortunately I had to comment out the `jit` because of the pattern shown in the comment led to a lot of recompilation because it was considered static.

My plan was to work around that case by recognizing that pattern outside the call to `_gather` and removing the specific indices from the static part of the index. Note this is a little subtle because of cases like `np.arange(10)[5:15]` mean the shape may depend on the index value, but note we only support concrete indices here anyway.

I suspect another blocker here is that there is, subsequent to the `jit` attempt above, a fast-path to generate a `reshape` in some common cases that don't actually do any slicing. We must not break that because the lazy constant stuff depends on it working.

Reply:
(We can always revise the lazy constant stuff if needed! All options are on the table.)

#2878 is kind of related, except it's about wanting to do static index munging (like on Python integers or integer onp.ndarrays) on the host instead of on device.

Reply:
I'm going to abandon this change. My ultimate goal is to address https://github.com/google/jax/issues/2871. After discussing with Peter, we think the easiest path forward for #2871 is to redo the sharding logic to use lax.slice instead of falling back to this code path like I did in https://github.com/google/jax/commit/07571ae4dd3fceee580aa49c4490f99ce7f6b6de. It would still be useful to speed up this path, but it's tricky as already discussed. I filed https://github.com/google/jax/issues/2891 to track this (I couldn't find an existing issue but maybe I missed it).